### ISSUE TITLE
[BUG] Service Alias Collision Error

### ISSUE THREAD

**Issue by Aran30** [rwth-acis/las2peer#109]
1. **Summary** - It sometimes happens that after restarting a service, it won't start anymore due to a service alias collision error. It seems that some metadata envelopes are fetched, but the actual service envelope is not found, and when wanting to upload the new service, it leads to a collision. 
2. **Bug Details** 
   1. **What?** - Network thinks service is already running and does not let service get started. 
   2. **Where?** - Servicestart in cluster, using ethereum and shared storage. 
   3. **When?/How often?** - Not that often. Suspicion that this happens when one node, which is connected to the network, has no PVC, which then leads to inconsistencies.

**Comment by Aran30** [comment]
The logs visible from the pod logs. 

```
Handling: 'startService('i5.las2peer.services.socialBotManagerService.SocialBotManagerService@1.0.19', 'sbf')'
Storing envelope agent-4dd661b9e5b86d5a681a831bbdc8c7ac1a85b49de24cc1432fa766868320531c21f801a4eebfeebc6c8bc3b688e947551a8350cd5fa0e91706afc84ed57c4a15#1 ...
ServiceAgent was not known in network. Published it
Aug 03, 2021 2:46:53 PM org.glassfish.jersey.internal.Errors logErrors
WARNING: The following warnings have been detected: WARNING: A HTTP GET method, public javax.ws.rs.core.Response i5.las2peer.services.socialBotManagerService.SocialBotManagerService.trainAndLoadStatus(java.lang.String), should not consume any entity.
MySQL root password SBF mobsos-mysql.mobsos 3306
Error while announcing deployment: i5.las2peer.registry.exceptions.EthereumException: Failed to submit service deployment announcement (DEBUG: SocialBotManagerService, java.lang.RuntimeException: Error processing transaction request: insufficient funds for gas * price + value)
Starting service failed
i5.las2peer.api.security.AgentException: Service instance problems!
	at i5.las2peer.security.ServiceAgentImpl.notifyRegistrationTo(ServiceAgentImpl.java:404)
	at i5.las2peer.p2p.Node.registerReceiver(Node.java:659)
	at i5.las2peer.p2p.PastryNodeImpl.registerReceiver(PastryNodeImpl.java:384)
	at i5.las2peer.tools.L2pNodeLauncher.startService(L2pNodeLauncher.java:704)
	at i5.las2peer.tools.L2pNodeLauncher.startServiceXml(L2pNodeLauncher.java:678)
	at i5.las2peer.tools.L2pNodeLauncher.startService(L2pNodeLauncher.java:653)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at i5.las2peer.tools.CommandPrompt.executeMethod(CommandPrompt.java:352)
	at i5.las2peer.tools.CommandPrompt.handleCommand(CommandPrompt.java:300)
	at i5.las2peer.tools.CommandPrompt.handleLine(CommandPrompt.java:459)
	at i5.las2peer.tools.L2pNodeLauncher.launchConfiguration(L2pNodeLauncher.java:1020)
	at i5.las2peer.tools.L2pNodeLauncher.main(L2pNodeLauncher.java:1185)
Caused by: i5.las2peer.api.ServiceException: Service alias is already used by another service!
	at i5.las2peer.security.ServiceAgentImpl.notifyRegistrationTo(ServiceAgentImpl.java:387)
	... 14 more
Caused by: i5.las2peer.p2p.AliasConflictException: Storage error.
	at i5.las2peer.p2p.ServiceAliasManager.registerServiceAlias(ServiceAliasManager.java:128)
	at i5.las2peer.security.ServiceAgentImpl.notifyRegistrationTo(ServiceAgentImpl.java:383)
	... 14 more
Caused by: i5.las2peer.api.persistency.EnvelopeNotFoundException: Part (0) of 'SERVICE_ALIAS-sbfmanager' with id (9EB12C2B796000B83735A6BEF75E397E1DE72312) not found in shared storage!
	at i5.las2peer.persistency.SharedStorage$7.onLookup(SharedStorage.java:541)
	at i5.las2peer.persistency.pastry.PastLookupContinuation$1.run(PastLookupContinuation.java:46)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
  --> Exception in executed method!
      InvocationTargetException (Service instance problems!)
      (print StackTrace with pst / printStackTrace)
```

**Comment by Aran30** [comment]
Some logs from inside the envelope message exchange logs:
las2peer.log.8:Caused by: i5.las2peer.api.ServiceException: Service alias is already used by another service!
```
bash-5.0$ grep 'error' *
las2peer.log.8:2021 Aug 03 09:04:15 FINER i5.las2peer.p2p.EthereumNode: ARTIFACT_UPLOAD_FAILED (-2200)  <0x608C63..>//137.226.232.175:31028:onyx-data-proxy-587f6bc588-lr4jt/10.244.0.81:31028      -       -       -       Storage error for Artifact SERVICE_ALIAS-onyx
las2peer.log.8:Caused by: i5.las2peer.p2p.AliasConflictException: Storage error.
bash-5.0$ grep 'SERVICE_ALIAS-onyx' *
las2peer.log.8:2021 Aug 03 09:04:15 FINER i5.las2peer.p2p.EthereumNode: ARTIFACT_FETCH_STARTED (2060)   <0x608C63..>//137.226.232.175:31028:onyx-data-proxy-587f6bc588-lr4jt/10.244.0.81:31028      -       -       -       SERVICE_ALIAS-onyx
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.SharedStorage: Starting latest version lookup for SERVICE_ALIAS-onyx at 1
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.helper.LatestArtifactVersionFinder: Looking for metadata envelope with identifier 'SERVICE_ALIAS-onyx' and version 1 at id 46F3F9E08934FECE6C49C72A2353BB2807EF22B3 ...
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.helper.LatestArtifactVersionFinder: Lookup got 3 past handles for identifier 'SERVICE_ALIAS-onyx' and version 1
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.helper.LatestArtifactVersionFinder: Looking for metadata envelope with identifier 'SERVICE_ALIAS-onyx' and version 2 at id 9E1E99490E7E27D798FC0846A2A005A75CBE8856 ...
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.helper.LatestArtifactVersionFinder: Lookup got 0 past handles for identifier 'SERVICE_ALIAS-onyx' and version 2
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.helper.LatestArtifactVersionFinder: found latest version (1) for SERVICE_ALIAS-onyx
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.SharedStorage: Fetching part (0) of envelope 'SERVICE_ALIAS-onyx' with id 7AC1DBCDC86C19BA16B0A4DECB551251940732E9 ...
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.SharedStorage: Got 0 past handles for part (0) of 'SERVICE_ALIAS-onyx'
las2peer.log.8:2021 Aug 03 09:04:15 FINER i5.las2peer.p2p.EthereumNode: ARTIFACT_FETCH_FAILED (-2065)   <0x608C63..>//137.226.232.175:31028:onyx-data-proxy-587f6bc588-lr4jt/10.244.0.81:31028      -       -       -       SERVICE_ALIAS-onyx
las2peer.log.8:2021 Aug 03 09:04:15 INFO i5.las2peer.persistency.SharedStorage: Storing envelope SERVICE_ALIAS-onyx#1 ...
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.SharedStorage: Checking collision for SERVICE_ALIAS-onyx#1
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.SharedStorage: Fetching part (0) of envelope 'SERVICE_ALIAS-onyx' with id 7AC1DBCDC86C19BA16B0A4DECB551251940732E9 ...
las2peer.log.8:2021 Aug 03 09:04:15 FINE i5.las2peer.persistency.SharedStorage: Got 0 past handles for part (0) of 'SERVICE_ALIAS-onyx'
las2peer.log.8:2021 Aug 03 09:04:15 FINER i5.las2peer.p2p.EthereumNode: ARTIFACT_UPLOAD_FAILED (-2200)  <0x608C63..>//137.226.232.175:31028:onyx-data-proxy-587f6bc588-lr4jt/10.244.0.81:31028      -       -       -       Storage error for Artifact SERVICE_ALIAS-onyx
las2peer.log.8:Caused by: i5.las2peer.api.persistency.EnvelopeNotFoundException: Part (0) of 'SERVICE_ALIAS-onyx' with id (7AC1DBCDC86C19BA16B0A4DECB551251940732E9) not found in shared storage!
```

**Comment by Navidda** [comment]
#### It happens also on MOBSOS Surveys
<details>
  <summary>Logs</summary>
  
```
Handling: 'startService('i5.las2peer.services.mobsos.surveys.SurveyService@0.3.1')'
Storing envelope agent-cb927998d51528fb3e9617b9bdc461b42297f05d10a58983fd374dee274aaa4e2e2d48e1176ee8b254d3f18570b8d063eade3317b7d21375db862fddc3432afe#1 ...
ServiceAgent was not known in network. Published it
Error while announcing deployment: i5.las2peer.registry.exceptions.EthereumException: Failed to submit service deployment announcement (DEBUG: SurveyService, java.lang.RuntimeException: Error processing transaction request: insufficient funds for gas * price + value)
	--> registering topic <0x527E58..>)
Successfully subscribed to topic '<0x527E58..>'
Storing envelope SERVICE_ALIAS-mobsos-surveys#1 ...
Error while announcing deployment: i5.las2peer.registry.exceptions.EthereumException: Failed to submit service deployment announcement (DEBUG: SurveyService, java.lang.RuntimeException: Error processing transaction request: insufficient funds for gas * price + value)
0xDC256E:rice.p2p.past.messaging@i5.las2peer.enterprise.storage:20210803.161200.410:ERROR: Outgoing PAST message [InsertMessage for JSPC [6AA4326B17035B48824E7389ED86E6168BC6CD29]] with UID -2147483576 was lost rice.p2p.past.PastException: Outgoing message '[InsertMessage for JSPC [6AA4326B17035B48824E7389ED86E6168BC6CD29]]' to null/[RSNH: <0x7379DA..>//137.226.232.175:31023:/10.244.0.82:31023] was lost - please try again.
	at rice.p2p.past.messaging.MessageLostMessage.returnResponse(MessageLostMessage.java:105)
	at rice.p2p.past.PastImpl.handleResponse(PastImpl.java:585)
	at rice.p2p.past.PastImpl.deliver(PastImpl.java:1015)
	at rice.pastry.commonapi.PastryEndpoint.messageForAppl(PastryEndpoint.java:502)
	at rice.pastry.commonapi.PastryEndpoint.receiveMessage(PastryEndpoint.java:594)
	at rice.pastry.messaging.MessageDispatch.dispatchMessage(MessageDispatch.java:156)
	at rice.pastry.PastryNode.receiveMessage(PastryNode.java:472)
	at rice.pastry.ScheduledMessage.run(ScheduledMessage.java:87)
	at rice.selector.TimerTask.execute(TimerTask.java:79)
	at rice.selector.SelectorManager.executeTask(SelectorManager.java:761)
	at rice.selector.SelectorManager.executeDueTasks(SelectorManager.java:727)
	at rice.selector.SelectorManager.run(SelectorManager.java:262)
...

...
0xDC256E:rice.p2p.past@i5.las2peer.enterprise.storage:20210803.161200.423:result[4]: rice.p2p.past.PastException: Outgoing message '[InsertMessage for JSPC [6AA4326B17035B48824E7389ED86E6168BC6CD29]]' to null/[RSNH: <0x199900..>//137.226.232.175:31012:/10.244.0.77:31012] was lost - please try again.
	at rice.p2p.past.messaging.MessageLostMessage.returnResponse(MessageLostMessage.java:105)
	at rice.p2p.past.PastImpl.handleResponse(PastImpl.java:585)
	at rice.p2p.past.PastImpl.deliver(PastImpl.java:1015)
	at rice.pastry.commonapi.PastryEndpoint.messageForAppl(PastryEndpoint.java:502)
	at rice.pastry.commonapi.PastryEndpoint.receiveMessage(PastryEndpoint.java:594)
	at rice.pastry.messaging.MessageDispatch.dispatchMessage(MessageDispatch.java:156)
	at rice.pastry.PastryNode.receiveMessage(PastryNode.java:472)
	at rice.pastry.ScheduledMessage.run(ScheduledMessage.java:87)
	at rice.selector.TimerTask.execute(TimerTask.java:79)
	at rice.selector.SelectorManager.executeTask(SelectorManager.java:761)
	at rice.selector.SelectorManager.executeDueTasks(SelectorManager.java:727)
	at rice.selector.SelectorManager.run(SelectorManager.java:262)
0xDC256E:rice.p2p.past@i5.las2peer.enterprise.storage:20210803.161200.423:result[5]: rice.p2p.past.PastException: Write access blocked! Different authors
	at i5.las2peer.persistency.EnvelopeArtifact.checkInsert(EnvelopeArtifact.java:25)
	at rice.p2p.past.PastImpl$11$1.receiveResult(PastImpl.java:1034)
	at rice.persistence.StorageManagerImpl$2.receiveResult(StorageManagerImpl.java:148)
	at rice.environment.processing.WorkRequest.returnResult(WorkRequest.java:70)
	at rice.environment.processing.WorkRequest$1.run(WorkRequest.java:87)
	at rice.selector.SelectorManager.doInvocations(SelectorManager.java:443)
	at rice.selector.SelectorManager.run(SelectorManager.java:264)
Monitoring: Problems with initializing Agents..
Starting service failed
i5.las2peer.api.security.AgentException: Service instance problems!
	at i5.las2peer.security.ServiceAgentImpl.notifyRegistrationTo(ServiceAgentImpl.java:404)
	at i5.las2peer.p2p.Node.registerReceiver(Node.java:659)
	at i5.las2peer.p2p.PastryNodeImpl.registerReceiver(PastryNodeImpl.java:384)
	at i5.las2peer.tools.L2pNodeLauncher.startService(L2pNodeLauncher.java:704)
	at i5.las2peer.tools.L2pNodeLauncher.startServiceXml(L2pNodeLauncher.java:678)
	at i5.las2peer.tools.L2pNodeLauncher.startService(L2pNodeLauncher.java:653)
	at i5.las2peer.tools.L2pNodeLauncher.startService(L2pNodeLauncher.java:606)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at i5.las2peer.tools.CommandPrompt.executeMethod(CommandPrompt.java:352)
	at i5.las2peer.tools.CommandPrompt.handleCommand(CommandPrompt.java:300)
	at i5.las2peer.tools.CommandPrompt.handleLine(CommandPrompt.java:459)
	at i5.las2peer.tools.L2pNodeLauncher.launchConfiguration(L2pNodeLauncher.java:1020)
	at i5.las2peer.tools.L2pNodeLauncher.main(L2pNodeLauncher.java:1185)
Caused by: i5.las2peer.api.ServiceException: Service alias is already used by another service!
	at i5.las2peer.security.ServiceAgentImpl.notifyRegistrationTo(ServiceAgentImpl.java:387)
	... 15 more
Caused by: i5.las2peer.p2p.AliasConflictException: Storage error.
	at i5.las2peer.p2p.ServiceAliasManager.registerServiceAlias(ServiceAliasManager.java:128)
	at i5.las2peer.security.ServiceAgentImpl.notifyRegistrationTo(ServiceAgentImpl.java:383)
	... 15 more
Caused by: i5.las2peer.api.persistency.EnvelopeException:
	at i5.las2peer.persistency.SharedStorage.waitForStoreResult(SharedStorage.java:154)
	at i5.las2peer.persistency.SharedStorage.storeEnvelope(SharedStorage.java:209)
	at i5.las2peer.p2p.PastryNodeImpl.storeEnvelope(PastryNodeImpl.java:681)
	at i5.las2peer.p2p.PastryNodeImpl.storeEnvelope(PastryNodeImpl.java:628)
	at i5.las2peer.p2p.ServiceAliasManager.createEntry(ServiceAliasManager.java:194)
	at i5.las2peer.p2p.ServiceAliasManager.registerServiceAlias(ServiceAliasManager.java:126)
	... 16 more
Caused by: rice.p2p.past.PastException: Had only 0 successful inserts out of 6 - aborting.
	at rice.p2p.past.PastImpl$6$1.isDone(PastImpl.java:681)
	at rice.Continuation$MultiContinuation.receive(Continuation.java:381)
	at rice.Continuation$MultiContinuation$1.receiveException(Continuation.java:363)
	at rice.Continuation$NamedContinuation.receiveException(Continuation.java:464)
	at rice.p2p.past.messaging.MessageLostMessage.returnResponse(MessageLostMessage.java:107)
	at rice.p2p.past.PastImpl.handleResponse(PastImpl.java:585)
	at rice.p2p.past.PastImpl.deliver(PastImpl.java:1015)
	at rice.pastry.commonapi.PastryEndpoint.messageForAppl(PastryEndpoint.java:502)
	at rice.pastry.commonapi.PastryEndpoint.receiveMessage(PastryEndpoint.java:594)
	at rice.pastry.messaging.MessageDispatch.dispatchMessage(MessageDispatch.java:156)
	at rice.pastry.PastryNode.receiveMessage(PastryNode.java:472)
	at rice.pastry.ScheduledMessage.run(ScheduledMessage.java:87)
	at rice.selector.TimerTask.execute(TimerTask.java:79)
	at rice.selector.SelectorManager.executeTask(SelectorManager.java:761)
	at rice.selector.SelectorManager.executeDueTasks(SelectorManager.java:727)
	at rice.selector.SelectorManager.run(SelectorManager.java:262)
  --> Exception in executed method!
      InvocationTargetException (Service instance problems!)
      (print StackTrace with pst / printStackTrace)
```
</details>

**Comment by Aran30** [comment]
I managed to reproduce the error locally with the following steps:

1. Started a first node and boostraped another one to it. 
2. Looked in the logs what the metadata and artifact envelope IDs are.
3. Delete the artifact envelope from the node storage.

This resulted in the error.
To fix it locally, I simply deleted the metadata envelope from the node storage.
Yet, I can't seem to find it in the cluster node storage...
